dodiscover/constraint/_classes.py

```diff
@@ -0,0 +1,263 @@
+        for (node_i, node_j) in itertools.combinations(init_graph.nodes, 2):
+            if not init_graph.has_edge(node_i, node_j):
+                sep_set[node_i][node_j] = []
+                sep_set[node_j][node_i] = []
+
+        return sep_set
+
+    def convert_skeleton_graph(self, graph: nx.Graph) -> EquivalenceClassProtocol:
+        raise NotImplementedError(
+            "All constraint discovery algorithms need to implement a function to convert "
+            "the skeleton graph to a causal graph."
+        )
+
+    def orient_unshielded_triples(
+        self, graph: EquivalenceClassProtocol, sep_set: Dict[str, Dict[str, List[Set[Any]]]]
+    ) -> None:
+        raise NotImplementedError()
+
+    def orient_edges(self, graph: EquivalenceClassProtocol) -> None:
+        raise NotImplementedError(
+            "All constraint discovery algorithms need to implement a function to orient the "
+            "skeleton graph given a separating set."
+        )
+
+    def fit(self, context: Context) -> None:
+        """Fit constraint-based discovery algorithm on dataset 'X'.
+
+        Parameters
+        ----------
+        X : Union[pd.DataFrame, Dict[Set, pd.DataFrame]]
+            Either a pandas dataframe constituting the endogenous (observed) variables
+            as columns and samples as rows, or a dictionary of different sampled
+            distributions with keys as the distribution names and values as the dataset
+            as a pandas dataframe.
+
+        Raises
+        ------
+        RuntimeError
+            If 'X' is a dictionary, then all datasets should have the same set of column
+            names (nodes).
+
+        Notes
+        -----
+        Control over the constraints imposed by the algorithm can be passed into the class
+        constructor.
+        """
+        self.context_ = context
+        graph = context.init_graph
+        self.init_graph_ = graph
+        self.fixed_edges_ = context.included_edges
+
+        # create a reference to the underlying data to be used
+        self.X_ = context.data
+
+        # initialize graph object to apply learning
+        sep_set = self._initialize_sep_sets(self.init_graph_)
+
+        # learn skeleton graph and the separating sets per variable
+        graph, sep_set = self.learn_skeleton(context, sep_set)
+
+        # convert networkx.Graph to relevant causal graph object
+        graph = self.convert_skeleton_graph(graph)
+
+        # orient edges on the causal graph object
+        if self.apply_orientations:
+            # for all pairs of non-adjacent variables with a common neighbor
+            # check if we can orient the edge as a collider
+            self.orient_unshielded_triples(graph, sep_set)
+            self.orient_edges(graph)
+
+        # store resulting data structures
+        self.separating_sets_ = sep_set
+        self.graph_ = graph
+
+    def test_edge(self, data: pd.DataFrame, X, Y, Z: Optional[Set] = None) -> Tuple[float, float]:
```

Comment:
Since ```test_``` is typically used for unit tests, this could be confusing. What about something like ```evaluate``` or ```verify``` (not sure which would be a better fit).